homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,173 @@
+from collections import defaultdict
+from datetime import datetime
+import logging
+
+from subarulink.hass import HassController as SubaruAPI, SubaruException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_BATTERY_LEVEL,
+    CONF_DEVICE_ID,
+    CONF_PASSWORD,
+    CONF_PIN,
+    CONF_SCAN_INTERVAL,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import aiohttp_client
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import slugify
+
+from .const import (
+    CONF_HARD_POLL_INTERVAL,
+    DEFAULT_HARD_POLL_INTERVAL,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    ICONS,
+    SUBARU_COMPONENTS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass, base_config):
+    """Set up of Subaru component."""
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Subaru from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    config = entry.data
+    websession = aiohttp_client.async_get_clientsession(hass)
+    date = datetime.now().strftime("%Y-%m-%d")
+    device_name = "Home Assistant: Added " + date
+    try:
+        controller = SubaruAPI(
+            websession,
+            config[CONF_USERNAME],
+            config[CONF_PASSWORD],
+            config[CONF_DEVICE_ID],
+            config[CONF_PIN],
+            device_name,
+            update_interval=entry.options.get(
+                CONF_HARD_POLL_INTERVAL, DEFAULT_HARD_POLL_INTERVAL
+            ),
+            fetch_interval=entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL),
+        )
+        if not await controller.connect():
+            _LOGGER.error("Failed to connect")
+            return False
+    except SubaruException as ex:
+        _LOGGER.error("Unable to communicate with Subaru API: %s", ex.message)
+        return False
+    entry_data = hass.data[DOMAIN][entry.entry_id] = {
+        "controller": controller,
+        "devices": defaultdict(list),
+        "listener": [entry.add_update_listener(update_listener)],
+    }
+
+    all_devices = entry_data["controller"].get_homeassistant_components()
+
+    if not all_devices:
+        return False
+
+    for device in all_devices:
+        entry_data["devices"][device.hass_type].append(device)
+
+    for component in SUBARU_COMPONENTS:
+        _LOGGER.debug("Loading %s", component)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in SUBARU_COMPONENTS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
```

Comment:
Should the DOMAIN be removed if all entries are removed?

Reply:
Reverted. DOMAIN is no longer removed since it is assumed to be in place by async_setup().  Otherwise if an entry is deleted and a new one is added, the DOMAIN won't be there.